File: Client/RelaxKonOS.Client.Android/app/src/main/java/app/relaxkonos/mobile/data/UploadResumeJournal.kt
package app.relaxkonos.mobile.data

import java.io.File

/**
 * One unfinished upload, remembered so a restart, a crash, a network change or a process kill continues
 * from the server's offset instead of re-sending a file that may already be nearly complete.
 *
 * It holds no file content and no credential: only the identifiers needed to ask the server where to
 * continue, plus the URI to read the bytes from again. That is why the file it is stored in is not
 * encrypted — there is nothing in it worth protecting, and saying so here keeps a later reader from
 * assuming the opposite.
 */
data class UploadResumeEntry(
    /** Server / user / workspace / device the session belongs to. An entry is never used against another. */
    val serverKey: String,
    val uploadId: String,
    val targetDirectoryPath: String,
    val fileName: String,
    /** The document this upload read from, kept so the bytes can be re-opened after a restart. */
    val sourceUri: String,
    /** The source signature the session was opened against; a changed document is not resumed. */
    val sourceKey: String,
    val confirmedOffset: Long,
    val totalLength: Long,
    val recordedAtMillis: Long,
)

/**
 * Device-local record of unfinished uploads.
 *
 * Stored as one escaped record per line rather than JSON, for one concrete reason: the Android unit
 * tests run against a stub `org.json` that throws on every call, so a JSON journal could only ever be
 * tested on a device. The format here is small enough to be exact and is exercised by a JVM test.
 */
class UploadResumeJournal(
    private val file: File,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    companion object {
        private const val HEADER = "relaxkonos-upload-resume 1"
        private const val FIELD_COUNT = 9

        /** Matches the server's absolute session lifetime: an older entry could not be resumed anyway. */
        const val MAXIMUM_AGE_MILLIS: Long = 7L * 24 * 60 * 60 * 1000

        /** Mobile devices hold fewer unfinished transfers than a desktop does. */
        const val MAXIMUM_ENTRIES = 16
    }

    /**
     * Finds a resumable entry for this exact document on this exact server.
     *
     * An entry whose signature no longer matches is removed rather than ignored: it can never be resumed
     * again, and leaving it would keep its cache copy alive forever.
     */
    fun find(serverKey: String, sourceKey: String, totalLength: Long): UploadResumeEntry? {
        val entries = read()
        val match = entries.lastOrNull {
            it.serverKey == serverKey && it.sourceKey == sourceKey && it.totalLength == totalLength
        }
        val stale = entries.filter {
            it.sourceKey == sourceKey && it.serverKey == serverKey && it.uploadId != match?.uploadId
        }
        if (stale.isNotEmpty()) {
            write(entries.filterNot { entry -> stale.any { it.uploadId == entry.uploadId } })
        }
        return match
    }

    /** Every unfinished entry for the current server, newest first. */
    fun entries(serverKey: String): List<UploadResumeEntry> =
        read().filter { it.serverKey == serverKey }.sortedByDescending { it.recordedAtMillis }

    fun record(entry: UploadResumeEntry) {
        // Read-modify-write on an immutable list: re-recording the same session replaces its entry rather
        // than appending a second one, which is what keeps "unfinished uploads" from growing duplicates.
        write(read().filterNot { it.uploadId == entry.uploadId } + entry)
    }

    /** Removes one entry and returns it, so its cache copy can be deleted by the same call site. */
    fun remove(uploadId: String): UploadResumeEntry? {
        val entries = read()
        val removed = entries.firstOrNull { it.uploadId == uploadId }
        if (removed != null) write(entries.filterNot { it.uploadId == uploadId })
        return removed
    }

    /**
     * Drops every entry that does not belong to [serverKey] and returns them.
     *
     * Called when the signed-in server / user / workspace / device changes: a session id is only
     * meaningful to the identity that opened it, so an entry from another one can never be resumed. The
     * removed entries are returned rather than forgotten because their cache copies have to go with them.
     */
    fun keepServer(serverKey: String): List<UploadResumeEntry> {
        val entries = read()
        val kept = entries.filter { it.serverKey == serverKey }
        if (kept.size != entries.size) write(kept)
        return entries.filterNot { it.serverKey == serverKey }
    }

    private fun read(): MutableList<UploadResumeEntry> {
        val lines = runCatching { file.readLines() }.getOrNull() ?: return mutableListOf()
        if (lines.firstOrNull()?.trim() != HEADER) return mutableListOf()
        val entries = mutableListOf<UploadResumeEntry>()
        for (line in lines.drop(1)) {
            if (line.isBlank()) continue
            parse(line)?.let { entries.add(it) }
        }
        return trim(entries)
    }

    private fun write(entries: List<UploadResumeEntry>) {
        val trimmed = trim(entries.toMutableList())
        runCatching {
            file.parentFile?.mkdirs()
            file.writeText(buildString {
                append(HEADER).append('\n')
                for (entry in trimmed) {
                    append(encode(entry)).append('\n')
                }
            })
        }
    }

    /**
     * Drops entries that cannot be resumed because they are older than the server would keep them, and
     * the oldest beyond the count limit, so the file cannot grow without bound.
     */
    private fun trim(entries: MutableList<UploadResumeEntry>): MutableList<UploadResumeEntry> {
        val cutoff = clock() - MAXIMUM_AGE_MILLIS
        entries.removeAll { it.recordedAtMillis < cutoff }
        if (entries.size > MAXIMUM_ENTRIES) {
            entries.sortBy { it.recordedAtMillis }
            while (entries.size > MAXIMUM_ENTRIES) entries.removeAt(0)
        }
        return entries
    }

    private fun encode(entry: UploadResumeEntry): String = listOf(
        entry.recordedAtMillis.toString(),
        entry.serverKey,
        entry.uploadId,
        entry.targetDirectoryPath,
        entry.fileName,
        entry.sourceUri,
        entry.sourceKey,
        entry.confirmedOffset.toString(),
        entry.totalLength.toString(),
    ).joinToString("\t") { escape(it) }

    private fun parse(line: String): UploadResumeEntry? = runCatching {
        val fields = line.split('\t')
        if (fields.size != FIELD_COUNT) return null
        val values = fields.map { unescape(it) }
        UploadResumeEntry(
            serverKey = values[1],
            uploadId = values[2],
            targetDirectoryPath = values[3],
            fileName = values[4],
            sourceUri = values[5],
            sourceKey = values[6],
            confirmedOffset = values[7].toLong(),
            totalLength = values[8].toLong(),
            recordedAtMillis = values[0].toLong(),
        ).takeIf { it.uploadId.isNotBlank() }
    }.getOrNull()

    private fun escape(value: String): String = buildString(value.length) {
        for (char in value) {
            when (char) {
                '\\' -> append("\\\\")
                '\t' -> append("\\t")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                else -> append(char)
            }
        }
    }

    /** Deliberately lenient: an unknown escape is kept verbatim rather than throwing the record away. */
    private fun unescape(value: String): String {
        if ('\\' !in value) return value
        val result = StringBuilder(value.length)
        var index = 0
        while (index < value.length) {
            val char = value[index]
            if (char != '\\' || index == value.length - 1) {
                result.append(char)
                index++
                continue
            }
            when (val next = value[index + 1]) {
                '\\' -> result.append('\\')
                't' -> result.append('\t')
                'n' -> result.append('\n')
                'r' -> result.append('\r')
                else -> result.append('\\').append(next)
            }
            index += 2
        }
        return result.toString()
    }
}

File: Client/RelaxKonOS.Client/Apps/Explorer/Uploads/UploadResumeJournal.cs
using System.Text.Json;

namespace RelaxKonOS.Client.Apps.Explorer.Uploads;

/// <summary>
/// One unfinished upload, remembered so a restart, a crash, or a lost response continues from the
/// server's offset instead of re-sending a file that may already be half transferred.
/// </summary>
/// <param name="UploadId">Server-assigned session identifier.</param>
/// <param name="ServerKey">Server / user / workspace / device the session belongs to. A journal entry is
/// never used against a different one: resuming into another server's session is resuming into someone
/// else's file.</param>
/// <param name="SourceLength">Length of the local file when the session was opened.</param>
/// <param name="SourceLastWriteUtc">Last write time of the local file when the session was opened.</param>
public sealed record UploadJournalEntry(
    string UploadId,
    string ServerKey,
    string TargetDirectoryPath,
    string FileName,
    string SourcePath,
    long SourceLength,
    DateTimeOffset SourceLastWriteUtc,
    DateTimeOffset RecordedAt);

/// <summary>
/// Device-local record of unfinished uploads. It holds no file content and no credential, only the
/// identifiers needed to ask the server where to continue.
/// </summary>
public sealed class UploadResumeJournal
{
    /// <summary>Matches the server's absolute session lifetime; an older entry cannot be resumed anyway.</summary>
    public static readonly TimeSpan MaximumAge = TimeSpan.FromDays(7);

    /// <summary>Oldest entries are dropped beyond this count, so the file cannot grow without bound.</summary>
    public const int MaximumEntries = 64;

    private readonly string _path;
    private readonly object _gate = new();
    private readonly List<UploadJournalEntry> _entries = [];
    private bool _loaded;

    public UploadResumeJournal(string? path = null)
    {
        _path = path ?? Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "RelaxKonOS", "upload-resume.json");
    }

    /// <summary>Path of the journal file, so a test can point it somewhere disposable.</summary>
    public string JournalPath => _path;

    /// <summary>
    /// Finds an entry that can be resumed for this exact source file on this exact server. A local file
    /// whose length or timestamp changed is not the file the session started with, so resuming would
    /// produce a file that is neither the old nor the new one; such an entry is discarded instead.
    /// </summary>
    public UploadJournalEntry? Find(string serverKey, string sourcePath, long length, DateTimeOffset lastWriteUtc)
    {
        lock (_gate)
        {
            EnsureLoaded();
            for (var index = _entries.Count - 1; index >= 0; index--)
            {
                var entry = _entries[index];
                if (!string.Equals(entry.ServerKey, serverKey, StringComparison.Ordinal)
                    || !string.Equals(entry.SourcePath, sourcePath, StringComparison.Ordinal)) continue;
                if (entry.SourceLength != length || entry.SourceLastWriteUtc != lastWriteUtc)
                {
                    _entries.RemoveAt(index);
                    Save();
                    continue;
                }
                return entry;
            }
            return null;
        }
    }

    public void Record(UploadJournalEntry entry)
    {
        lock (_gate)
        {
            EnsureLoaded();
            // One entry per source-and-destination. The upload id alone would not be enough: a session the
            // server had forgotten is replaced by a new one with a new id, and leaving its predecessor behind
            // means a later resume can pick a session that no longer exists.
            _entries.RemoveAll(existing =>
                string.Equals(existing.UploadId, entry.UploadId, StringComparison.Ordinal)
                || (string.Equals(existing.ServerKey, entry.ServerKey, StringComparison.Ordinal)
                    && string.Equals(existing.SourcePath, entry.SourcePath, StringComparison.Ordinal)
                    && string.Equals(existing.TargetDirectoryPath, entry.TargetDirectoryPath, StringComparison.Ordinal)
                    && string.Equals(existing.FileName, entry.FileName, StringComparison.Ordinal)));
            _entries.Add(entry);
            Trim();
            Save();
        }
    }

    public void Remove(string uploadId)
    {
        lock (_gate)
        {
            EnsureLoaded();
            if (_entries.RemoveAll(entry => string.Equals(entry.UploadId, uploadId, StringComparison.Ordinal)) > 0) Save();
        }
    }

    /// <summary>Drops every entry that belongs to a server key no longer in use. Called when the signed-in
    /// server/user/workspace/device changes, so a stale entry can never be offered to a new session.</summary>
    public void RemoveServer(string serverKey)
    {
        lock (_gate)
        {
            EnsureLoaded();
            if (_entries.RemoveAll(entry => !string.Equals(entry.ServerKey, serverKey, StringComparison.Ordinal)) > 0) Save();
        }
    }

    public IReadOnlyList<UploadJournalEntry> Snapshot()
    {
        lock (_gate)
        {
            EnsureLoaded();
            return [.. _entries];
        }
    }

    private void Trim()
    {
        var cutoff = DateTimeOffset.UtcNow - MaximumAge;
        _entries.RemoveAll(entry => entry.RecordedAt < cutoff);
        if (_entries.Count <= MaximumEntries) return;
        _entries.Sort((left, right) => left.RecordedAt.CompareTo(right.RecordedAt));
        _entries.RemoveRange(0, _entries.Count - MaximumEntries);
    }

    private void EnsureLoaded()
    {
        if (_loaded) return;
        _loaded = true;
        try
        {
            if (!File.Exists(_path)) return;
            var loaded = JsonSerializer.Deserialize<List<UploadJournalEntry>>(File.ReadAllText(_path));
            if (loaded is null) return;
            _entries.AddRange(loaded.Where(entry => !string.IsNullOrWhiteSpace(entry.UploadId)));
            Trim();
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException or JsonException)
        {
            // An unreadable journal means "no resumable uploads", which costs at most a retransmission.
            // Failing the upload because a bookkeeping file is corrupt would be far worse.
            _entries.Clear();
        }
    }

    private void Save()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
            File.WriteAllText(_path, JsonSerializer.Serialize(_entries));
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            // Journal writes are best effort for the same reason reads are: never fail the transfer over one.
        }
    }
}
